File: packages/shared/src/agent-runtimes.ts
import { z } from "zod";

import registry from "./agent-runtimes.json" with { type: "json" };

/**
 * The agent-runtime roster, loaded from `agent-runtimes.json`.
 *
 * That file is the *only* place a runtime is declared. The API's ingest path,
 * the desktop's labels and process folding, the capture probes and connection
 * instructions, and the quota dials all read it, so adding a runtime is a
 * configuration change rather than a code change. The Rust host reads the same
 * JSON (`agent_runtimes.rs` embeds it), so the two languages cannot drift.
 *
 * The roster is *not* an allowlist. `agentSourceSchema` accepts any id of the
 * canonical shape, so a runtime nobody has declared yet is still recorded under
 * its own name rather than collapsed into `other` or dropped. The roster only
 * decides what SIQshift can say *about* a runtime: its display name, its
 * executables, where its capture config lives, and which quota dial it backs.
 */

/** How SIQshift connects a runtime's lifecycle capture. */
export const agentRuntimeRegistrationValues = ["claude_json", "codex_native", "cursor_json", "manual"] as const;
export type AgentRuntimeRegistration = (typeof agentRuntimeRegistrationValues)[number];

/** Whether a runtime's capture mechanism names the model it is driving. */
export const agentRuntimeReportsModelValues = ["always", "sometimes", "never"] as const;
export type AgentRuntimeReportsModel = (typeof agentRuntimeReportsModelValues)[number];

const agentRuntimeSchema = z
  .object({
    /** Canonical source id, snake_case; what `agent_sessions.source` stores. */
    id: z.string().max(40).regex(/^[a-z][a-z0-9_]*$/),
    /** Display name, so one tool is never called two different things. */
    label: z.string().min(1),
    /** Executable basenames the runtime surfaces under in foreground-process stats. */
    binaries: z.array(z.string().min(1)).readonly(),
    /** The `quota-axi` provider backing this runtime, when one does. */
    quotaProvider: z.string().min(1).nullable(),
    /** Home-relative path to the runtime config SIQshift connects or inspects. */
    configPath: z.string().min(1),
    registration: z.enum(agentRuntimeRegistrationValues),
    /**
     * Whether the runtime's capture mechanism names the model it is driving:
     * `always` by design on every event, `sometimes` when it depends on the
     * user's wiring or an unconfirmed mechanism, `never` when the mechanism
     * cannot name one. Missing model evidence records none, never a guess.
     */
    reportsModel: z.enum(agentRuntimeReportsModelValues),
    /**
     * Id of a vendored official icon, or `null` when no official asset could be
     * sourced cleanly — those runtimes get the generic agent treatment rather
     * than a lookalike mark.
     */
    icon: z.string().min(1).nullable(),
    /** Paste-it-yourself lines for `manual` runtimes; `{command}` is the hook binary path. */
    manualSnippet: z.array(z.string()).readonly(),
  })
  .strict();

export type AgentRuntime = z.infer<typeof agentRuntimeSchema>;

/** A malformed roster is a build-time mistake, so it fails loudly at import. */
export const agentRuntimes: readonly AgentRuntime[] = z
  .array(agentRuntimeSchema)
  .readonly()
  .parse(registry.runtimes);

/** Canonical ids of the declared runtimes, in roster order. */
export const agentRuntimeIds: readonly string[] = agentRuntimes.map((runtime) => runtime.id);

const byId = new Map(agentRuntimes.map((runtime) => [runtime.id, runtime]));

export const findAgentRuntime = (id: string): AgentRuntime | undefined => byId.get(id);

/**
 * The name to show for a source. An undeclared runtime keeps its own id rather
 * than being relabelled into something the roster does know, because a wrong
 * name is worse than a raw one.
 */
export const agentRuntimeLabel = (id: string): string => byId.get(id)?.label ?? id;

const byBinary = new Map(
  agentRuntimes.flatMap((runtime) => runtime.binaries.map((binary) => [binary, runtime.id] as const)),
);

/**
 * The runtime an executable belongs to, for folding agent CLIs out of the
 * foreground-process rows. Matching is on the lowercased basename without a
 * `.exe` suffix; anything unrecognized is not an agent as far as this says.
 */
export const agentRuntimeForBinary = (processName: string): string | undefined =>
  byBinary.get(processName.replace(/\.exe$/i, "").toLowerCase());

/** The hook snippet for a `manual` runtime, with the hook binary path substituted in. */
export const agentRuntimeManualSnippet = (id: string, command: string): string | undefined => {
  const runtime = byId.get(id);
  if (runtime === undefined || runtime.manualSnippet.length === 0) return undefined;
  return runtime.manualSnippet.join("\n").replaceAll("{command}", command);
};

File: apps/desktop/src-tauri/src/agent_runtimes.rs
//! The agent-runtime roster.
//!
//! `packages/shared/src/agent-runtimes.json` is the single declaration of every
//! runtime SIQshift knows by name, and this module embeds that exact file so
//! the host and the TypeScript side cannot drift. Adding a runtime is an entry
//! in that JSON: the hook probes, the registration snippets, the display names,
//! the process folding, and the quota dials all read it.
//!
//! The roster is not an allowlist. `AgentSource` accepts any id of the
//! canonical shape, so a runtime nobody has declared yet is still spooled and
//! uploaded under its own name. The roster only decides what SIQshift can *say*
//! about a runtime.

use std::sync::OnceLock;

use serde::Deserialize;

/// Compiled in, so a build always ships a roster that matches its contracts.
const REGISTRY_JSON: &str = include_str!("../../../../packages/shared/src/agent-runtimes.json");

/// How SIQshift can switch a runtime's hooks on.
#[derive(Clone, Copy, Debug, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
pub enum Registration {
    /// Claude Code's nested `hooks.SessionStart[].hooks[]` arrays.
    ClaudeJson,
    /// Codex terminals discovered from held writer locks and app-server
    /// metadata. No lifecycle hook is installed.
    CodexNative,
    /// Cursor's flat `hooks.sessionStart[]` array of argv-carrying commands.
    CursorJson,
    /// No config shape the host will rewrite: the user gets an honest snippet.
    Manual,
}

/// The host's view of a roster entry. The roster carries more per runtime than
/// this — display labels, executables, quota providers, icons — but those are
/// the webview's business, so the host reads only what it acts on and ignores
/// the rest.
#[derive(Clone, Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AgentRuntime {
    pub id: String,
    /// Home-relative path to the config a hook registration lands in.
    pub config_path: String,
    pub registration: Registration,
    /// Paste-it-yourself lines; `{command}` stands in for the hook binary path.
    pub manual_snippet: Vec<String>,
}

#[derive(Debug, Deserialize)]
struct Registry {
    runtimes: Vec<AgentRuntime>,
}

/// The declared runtimes, in roster order. A malformed roster is a build-time
/// mistake, so it panics rather than silently shipping a shorter list.
pub fn runtimes() -> &'static [AgentRuntime] {
    static REGISTRY: OnceLock<Vec<AgentRuntime>> = OnceLock::new();
    REGISTRY
        .get_or_init(|| {
            serde_json::from_str::<Registry>(REGISTRY_JSON)
                .expect("the bundled agent-runtime roster parses")
                .runtimes
        })
        .as_slice()
}

pub fn find(id: &str) -> Option<&'static AgentRuntime> {
    runtimes().iter().find(|runtime| runtime.id == id)
}

/// The paste-it-yourself text for a runtime the host will not rewrite.
pub fn manual_snippet(id: &str, command: &str) -> Option<String> {
    let runtime = find(id)?;
    if runtime.manual_snippet.is_empty() {
        return None;
    }
    Some(
        runtime
            .manual_snippet
            .join("\n")
            .replace("{command}", command),
    )
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn the_bundled_roster_parses_and_declares_the_runtimes_in_use() {
        let ids: Vec<&str> = runtimes()
            .iter()
            .map(|runtime| runtime.id.as_str())
            .collect();
        for expected in [
            "claude_code",
            "codex",
            "cursor",
            "kimi_code",
            "pi",
            "opencode",
        ] {
            assert!(
                ids.contains(&expected),
                "{expected} is missing from the roster"
            );
        }
        assert_eq!(
            find("codex").map(|runtime| runtime.registration),
            Some(Registration::CodexNative),
            "Codex terminals come from the runtime's native inventory, not Claude-shaped hooks",
        );
    }

    #[test]
    fn runtimes_not_installed_anywhere_are_still_declared() {
        // The roster is a declaration, not a survey of this machine: absence is
        // never hard-coded, so these resolve without any of them being present.
        for id in ["pi_signed", "grok", "muse", "copilot"] {
            assert!(find(id).is_some(), "{id} is missing from the roster");
        }
    }

    #[test]
    fn an_undeclared_runtime_resolves_to_nothing_rather_than_to_a_neighbour() {
        assert!(find("some_new_agent").is_none());
        assert!(manual_snippet("some_new_agent", "/opt/siqshift-hook").is_none());
    }

    #[test]
    fn every_manual_runtime_carries_a_snippet_and_no_merged_one_does() {
        for runtime in runtimes() {
            let snippet = manual_snippet(&runtime.id, "/opt/siqshift-hook");
            match runtime.registration {
                Registration::Manual => {
                    let snippet = snippet.expect("a manual runtime explains itself");
                    assert!(
                        snippet.contains("/opt/siqshift-hook"),
                        "{} drops the command",
                        runtime.id
                    );
                    assert!(
                        !snippet.contains("{command}"),
                        "{} leaves a placeholder",
                        runtime.id
                    );
                }
                _ => assert!(
                    snippet.is_none(),
                    "{} merges, so it needs no snippet",
                    runtime.id
                ),
            }
        }
    }

    #[test]
    fn ids_are_unique_and_canonically_shaped() {
        let mut ids: Vec<&str> = runtimes()
            .iter()
            .map(|runtime| runtime.id.as_str())
            .collect();
        for id in &ids {
            assert_eq!(
                crate::spool::AgentSource::parse(id).map(|source| source.as_str().to_string()),
                Ok((*id).to_string()),
                "{id} is not already canonical",
            );
        }
        ids.sort_unstable();
        let count = ids.len();
        ids.dedup();
        assert_eq!(ids.len(), count, "two runtimes share an id");
    }
}
